addons/website_sale/controllers/main.py

```diff
@@ -605,9 +605,15 @@ def values_postprocess(self, order, mode, values, errors, error_msg):
         new_values['customer'] = True
         new_values['team_id'] = request.website.salesteam_id and request.website.salesteam_id.id
         new_values['user_id'] = request.website.salesperson_id and request.website.salesperson_id.id
-        new_values['company_id'] = request.website.company_id.id
         new_values['website_id'] = request.website.id
 
+        if order.partner_id.id == request.website.user_id.sudo().partner_id.id:
+            # If the partner is public, we assign the company of the website
+            new_values['company_id'] = request.website.company_id.id
+        else:
+            # Otherwise, we use the company of the connected partner
+            new_values['company_id'] = order.partner_id.company_id.id
```

Comment:
partner_id.company_id should match the one of the user. I don't really know which one to use here.